File: _examples/Kotlin/src/main/kotlin/com/coredf/examples/FullIntegrationStep.kt
// Copyright Core DF — Apache License 2.0
package com.coredf.examples

import com.coredf.cawbs.Cawbs
import com.coredf.files.FileClient
import com.coredf.kafka.KafkaClient
import com.coredf.transform.TransformClient

fun main() {
    check(Cawbs.init().statusCode == 200) { "cawbs.init failed" }
    val event = Cawbs.getEventPayload()
    check(event.statusCode == 200) { "getEventPayload failed" }

    val orderId = when (val p = event.payload) {
        is Map<*, *> -> (p["orderId"] ?: p["id"])?.toString() ?: "unknown"
        else -> "unknown"
    }
    val ackDir = System.getenv("EXAMPLE_ACK_DIR") ?: "/tmp/coreauto-example"
    val ackPath = "$ackDir/$orderId.json"
    val order = mapOf("orderId" to orderId, "details" to event.payload)
    val text = TransformClient.JsonStringify(order)
    check(text.statusCode == 200)
    check(FileClient.LocalWrite(ackPath, text.get("text").toString()).statusCode == 200)

    val topic = System.getenv("EXAMPLE_KAFKA_TOPIC") ?: "orders.enriched"
    val published = mutableListOf<String>()
    if (KafkaClient.Produce(topic, order).statusCode == 200) published.add("kafka")

    val out = mapOf("orderId" to orderId, "queuesPublished" to published, "ackPath" to ackPath)
    check(Cawbs.putStepPayload(out).statusCode == 200)
    println("""{"status_code":200,"result":$out}""")
}

File: _examples/Scala/src/main/scala/com/coredf/examples/FullIntegrationStep.scala
// Copyright Core DF — Apache License 2.0
package com.coredf.examples

import com.coredf.cawbs.Cawbs
import com.coredf.files.FileClient
import com.coredf.kafka.KafkaClient
import com.coredf.transform.TransformClient

@main def fullIntegrationStep(): Unit =
  require(Cawbs.Init().statusCode == 200, "cawbs.Init failed")
  val event = Cawbs.GetEventPayload()
  require(event.statusCode == 200, "GetEventPayload failed")

  val orderId = event.payload match
    case m: Map[?, ?] =>
      val keys = m.map { case (k, v) => k.toString -> v }
      keys.get("orderId").orElse(keys.get("id")).map(_.toString).getOrElse("unknown")
    case _ => "unknown"

  val ackDir = sys.env.getOrElse("EXAMPLE_ACK_DIR", "/tmp/coreauto-example")
  val ackPath = s"$ackDir/$orderId.json"
  val order = Map("orderId" -> orderId, "details" -> event.payload)
  val text = TransformClient.JsonStringify(order)
  require(text.statusCode == 200)
  require(FileClient.LocalWrite(ackPath, text.get("text").toString).statusCode == 200)

  val topic = sys.env.getOrElse("EXAMPLE_KAFKA_TOPIC", "orders.enriched")
  var published = List.empty[String]
  if KafkaClient.Produce(topic, order).statusCode == 200 then published = "kafka" :: published

  val out = Map("orderId" -> orderId, "queuesPublished" -> published.reverse, "ackPath" -> ackPath)
  require(Cawbs.PutStepPayload(out).statusCode == 200)
  println(s"""{"status_code":200,"result":$out}""")
